File: sdk/python/examples/basic_usage.py
"""
Basic usage example for RTDB Python SDK.

This example demonstrates:
- Connecting to RTDB server
- Creating collections
- Upserting vectors
- Searching for similar vectors
"""

import asyncio
import random

# Try native extension first, fallback to pure Python
try:
    from rtdb_sdk.rtdb_python import RtdbClient
    print("Using native Rust extension")
except ImportError:
    from rtdb_sdk.client import RtdbClient
    print("Using pure Python implementation")


async def main():
    # Connect to RTDB (default: http://localhost:6333)
    async with RtdbClient("http://localhost:6333") as client:
        
        # Check if server is healthy
        healthy = await client.is_healthy()
        print(f"Server healthy: {healthy}")
        
        if not healthy:
            print("Server is not responding. Make sure RTDB is running.")
            return
        
        # Create a collection for 128-dimensional vectors
        print("\nCreating collection 'documents'...")
        await client.create_collection(
            name="documents",
            dimension=128,
            distance="Cosine"  # or "Euclidean", "Dot"
        )
        print("Collection created!")
        
        # Generate and upsert some sample vectors
        print("\nUpserting sample vectors...")
        num_vectors = 100
        vectors = []
        
        for i in range(num_vectors):
            # Generate random normalized vector
            vec = [random.gauss(0, 1) for _ in range(128)]
            magnitude = sum(x**2 for x in vec) ** 0.5
            vec = [x / magnitude for x in vec]  # Normalize
            
            vectors.append({
                "id": i + 1,
                "vector": vec,
                "payload": {
                    "index": i,
                    "category": random.choice(["tech", "science", "art"]),
                    "score": random.random() * 100
                }
            })
        
        await client.upsert("documents", vectors)
        print(f"Upserted {num_vectors} vectors")
        
        # Search for similar vectors
        print("\nSearching for similar vectors...")
        query_vector = vectors[0]["vector"]  # Use first vector as query
        
        results = await client.search(
            collection_name="documents",
            vector=query_vector,
            limit=5,
            with_payload=True
        )
        
        print("\nTop 5 similar vectors:")
        for result in results:
            print(f"  ID: {result.id}, Score: {result.score:.4f}")
            if hasattr(result, 'payload') and result.payload:
                print(f"    Payload: {result.payload}")
        
        # Get a specific point
        print("\nRetrieving point ID 10...")
        point = await client.get_point("documents", 10)
        print(f"Point ID: {point.id}, Vector length: {len(point.vector)}")
        
        # Count points
        count = await client.count("documents")
        print(f"\nTotal points in collection: {count}")
        
        # List collections
        collections = await client.list_collections()
        print(f"\nCollections: {collections}")
        
        # Cleanup
        print("\nCleaning up...")
        await client.delete_collection("documents")
        print("Collection deleted!")


if __name__ == "__main__":
    asyncio.run(main())

File: sdk/javascript/examples/basic-usage.js
/**
 * Basic Usage Example for RTDB JavaScript SDK
 */

const { RTDBClient, FilterBuilder, PointBuilder, VectorUtils } = require('@rtdb/client');

async function basicExample() {
  // Create client
  const client = new RTDBClient({
    url: 'http://localhost:6333',
    apiKey: 'your-api-key' // Optional
  });

  try {
    // Check if server is ready
    const isReady = await client.isReady();
    console.log('Server ready:', isReady);

    // Create a collection
    const collectionName = 'my_collection';
    await client.createCollection(collectionName, {
      vector_size: 128,
      distance: 'Cosine'
    });
    console.log('Collection created');

    // Create some points
    const points = [
      new PointBuilder()
        .id('doc-1')
        .vector(VectorUtils.random(128))
        .addPayload('title', 'First document')
        .addPayload('category', 'tech')
        .build(),
      
      new PointBuilder()
        .id('doc-2')
        .vector(VectorUtils.random(128))
        .addPayload('title', 'Second document')
        .addPayload('category', 'science')
        .build(),
      
      new PointBuilder()
        .id('doc-3')
        .vector(VectorUtils.random(128))
        .addPayload('title', 'Third document')
        .addPayload('category', 'tech')
        .build()
    ];

    // Insert points
    const upsertResponse = await client.upsert(collectionName, { points });
    console.log('Points inserted:', upsertResponse.operation_id);

    // Wait a moment for indexing
    await new Promise(resolve => setTimeout(resolve, 1000));

    // Search for similar vectors
    const searchResponse = await client.search(collectionName, {
      vector: points[0].vector, // Use first point's vector as query
      limit: 2,
      with_payload: true
    });

    console.log('Search results:');
    searchResponse.result.forEach((result, index) => {
      console.log(`  ${index + 1}. ID: ${result.id}, Score: ${result.score.toFixed(4)}`);
      console.log(`     Title: ${result.payload?.title}`);
      console.log(`     Category: ${result.payload?.category}`);
    });

    // Search with filter
    const filter = new FilterBuilder()
      .equals('category', 'tech')
      .build();

    const filteredSearch = await client.search(collectionName, {
      vector: VectorUtils.random(128),
      limit: 10,
      filter,
      with_payload: true
    });

    console.log('\nFiltered search results (tech category):');
    filteredSearch.result.forEach((result, index) => {
      console.log(`  ${index + 1}. ID: ${result.id}, Title: ${result.payload?.title}`);
    });

    // Get a specific point
    const point = await client.getPoint(collectionName, 'doc-1', true, false);
    console.log('\nRetrieved point:', {
      id: point?.id,
      title: point?.payload?.title,
      hasVector: !!point?.vector
    });

    // Count points
    const totalCount = await client.count(collectionName);
    console.log('\nTotal points:', totalCount);

    const techCount = await client.count(collectionName, { filter });
    console.log('Tech category points:', techCount);

    // List collections
    const collections = await client.listCollections();
    console.log('\nAll collections:', collections);

    // Cleanup
    await client.deleteCollection(collectionName);
    console.log('Collection deleted');

  } catch (error) {
    console.error('Error:', error.message);
    if (error.status) {
      console.error('Status:', error.status);
    }
  } finally {
    // Always close the client
    client.close();
  }
}

// Run the example
if (require.main === module) {
  basicExample().catch(console.error);
}

module.exports = { basicExample };